pubnub/event_engine/effects.py

```diff
@@ -0,0 +1,431 @@
+        elif response.status.error:
+            self.logger.warning(f'Handshake failed: {response.status.error_data.__dict__}')
+            handshake_failure = events.HandshakeFailureEvent(response.status.error_data, 1, timetoken=timetoken)
+            self.event_engine.trigger(handshake_failure)
+        else:
+            cursor = response.result['t']
+            timetoken = timetoken if timetoken > 0 else cursor['t']
+            region = cursor['r']
+            handshake_success = events.HandshakeSuccessEvent(timetoken, region)
+            self.event_engine.trigger(handshake_success)
+
+
+class ReceiveMessagesEffect(Effect):
+    invocation: invocations.ReceiveMessagesInvocation
+
+    def run(self):
+        channels = self.invocation.channels
+        groups = self.invocation.groups
+        timetoken = self.invocation.timetoken
+        region = self.invocation.region
+
+        if hasattr(self.pubnub, 'event_loop'):
+            self.stop_event = self.get_new_stop_event()
+            self.run_async(self.receive_messages_async(channels, groups, timetoken, region))
+
+    async def receive_messages_async(self, channels, groups, timetoken, region):
+        request = Subscribe(self.pubnub)
+        if channels:
+            request.channels(channels)
+        if groups:
+            request.channel_groups(groups)
+        if timetoken:
+            request.timetoken(timetoken)
+        if region:
+            request.region(region)
+
+        request.cancellation_event(self.stop_event)
+        response = await request.future()
+
+        if response.status is None and response.result is None:
+            self.logger.warning('Recieve messages failed: Empty response')
+            recieve_failure = events.ReceiveFailureEvent('Empty response', 1, timetoken=timetoken)
+            self.event_engine.trigger(recieve_failure)
+        elif response.status.error:
+            self.logger.warning(f'Recieve messages failed: {response.status.error_data.__dict__}')
+            recieve_failure = events.ReceiveFailureEvent(response.status.error_data, 1, timetoken=timetoken)
+            self.event_engine.trigger(recieve_failure)
+        else:
+            cursor = response.result['t']
+            timetoken = cursor['t']
+            region = cursor['r']
+            messages = response.result['m']
+            recieve_success = events.ReceiveSuccessEvent(timetoken, region=region, messages=messages)
+            self.event_engine.trigger(recieve_success)
+        self.stop_event.set()
+
+
+class ReconnectEffect(Effect):
+    invocation: invocations.ReconnectInvocation
+    reconnection_policy: PNReconnectionPolicy
+
+    def __init__(self, pubnub_instance, event_engine_instance,
+                 invocation: Union[invocations.PNManageableInvocation, invocations.PNCancelInvocation]) -> None:
+        super().__init__(pubnub_instance, event_engine_instance, invocation)
+        self.reconnection_policy = pubnub_instance.config.reconnect_policy
```

Comment:
I've got an impression this could also be moved to the class or whatever type is responsible for computing reconnection delay. Think about it while working on a new reconnection policy.

Reply:
CLEN-1799